### ISSUE TITLE
Road to 1.0.0 

### ISSUE THREAD

**Issue by cgrindel** [cgrindel/rules_swift_package_manager#924]
## Goals

1. Remove need for Swift index JSON file.
2. Move Gazelle plugin to its own repository.
3. Create release 1.0.0.

## Tasks

- [x] Process SPM `Package.swift` and `Package.resolved` in the repository rule.
  - [x] Stop reading the index file in the repository rules.
  - [x] Be sure to return the correct repository info so that `mod tidy` will update the `MODULE.bazel` file with the correct `use_repos` declaration.
  - [x] Remove all of the index files in the examples.
  - [x] Remove all of the `swift_deps START` and `swift_deps END` comments.
  - [x] Update `bzlmod/workspace` to use `from_package()`.
- [x] Update the Gazelle plugin to not use the Swift index JSON file.
- [x] ~~Update `//:swift_update_pkgs` to just run SPM commands without using Go. Be sure to use the Swift toolchain, not the Swift installed on the system. (Possibly, create a new macro that creates the commands for running Swift package commands.)~~ For now, I removed `swift_update_packages`. If people would like a way to update Swift packages using Bazel, we will do it as a separate feature.
- [x] Fix examples
  - [x] `grpc_package_example`
  - [x] `symlink_example`
  - [x] `ios_sim`
- [x] Move Gazelle plugin to its own repository.
  - [x] Create a repository for the Gazelle plugin. https://github.com/cgrindel/swift_gazelle_plugin/pull/1
  - [x] Publish to BCR. https://github.com/bazelbuild/bazel-central-registry/pull/4059
  - [x] Remove Gazelle plugin from `rules_swift_package_manager`. #1538 
- [x] #1079 
- [x] #1168
- [x] #1563 
- [ ] Create release 1.0.0.

**Comment by cgrindel** [comment]
Notes regarding Gazelle plugin

- No more `update_repos`.
- `rules_swift_package_manager` will define a repository called `swift_deps_info`. It will contain the list of direct dependencies for the workspace (i.e., those defined in the `Package.swift`).
- The plugin 
  - It will use [repo.FindExternalRepo](https://pkg.go.dev/github.com/bazelbuild/bazel-gazelle@v0.36.0/repo#FindExternalRepo) to find the `swift_deps_info` and read the information written there. (NOTE: We could just read the list of repo names by parsing the `MODULE.bazel`. However, I think that we want to support non-bzlmod a bit longer.)
  - It will then read the `pkg_info.json` file that lives at the root of every Swift package repository.
  - This will populate the `DependencyIndex`.

**Comment by brentleyjones** [comment]
What do we need `DependencyIndex` for anymore? Aren't product names deterministic now?

**Comment by cgrindel** [comment]
> What do we need DependencyIndex for anymore? Aren't product names deterministic now?
The Gazelle plugin needs to resolve module names to Bazel labels.

**Comment by brentleyjones** [comment]
Ahh, so that won't be in this repo in the end? Got it.

**Comment by cgrindel** [comment]
For anyone following along at home, this branch has the examples working if you run `bazel test //...`, no index file required. 🙂 I started updating the Gazelle plugin to work properly in this new world. Here is the basic plan:

- Provide a `swift_deps_info` repository rule that generates a JSON file with the module/product/package information that is needed to resolve modules to Bazel targets.
- Update the Gazelle plugin to read this file and load it as the `DependencyIndex`.
- I anticipate keeping the existing API for the `DependencyIndex`. So, I hope that there won't be any other major changes to the resolve and generate portions of the Gazelle plugin.
- Remove the update repos portion of the Gazelle plugin.

**Comment by cgrindel** [comment]
> Can we add https://github.com/cgrindel/rules_swift_package_manager/issues/1079 to this?

Done.

**Comment by cgrindel** [comment]
Just a quick update regarding the v1 release. I have been working on #1079. I have a draft PR (#1217) with everything working except the `rules_xcodeproj` no-sandbox build. Unfortunately, that fails with duplicate definitions. I see and mostly understand why this is happening. Unfortunately, I am struggling to prevent the duplicate module map files causing the errors.

I put together #1266 which attempts to prevent `objc_library` from creating its own module map file by specifying the one that `rules_swift_package_manager` creates. Unfortunately, this causes an error:
```
In file included from external/rules_swift_package_manager~~swift_deps~swiftpkg_firebase_ios_sdk/FirebaseInstallations/Source/Library/FIRInstallations.m:25:
external/rules_swift_package_manager~~swift_deps~swiftpkg_firebase_ios_sdk/FirebaseCore/Extension/FirebaseCoreInternal.h:15:1: error: use of '@import' when modules are disabled
@import FirebaseCore;
^
1 error generated.
```
Looking through the actual command-line, I do not see the `-fmodules` flag. I thought that this was added by setting `enable_modules = True`. 

If anyone has any thoughts or suggestions, please send them my way.

**Comment by cgrindel** [comment]
The `cgrindel/swift_gazelle_plugin` has been [submitted to the BCR](https://github.com/bazelbuild/bazel-central-registry/pull/4059).

**Comment by cgrindel** [comment]
@brentleyjones @jpsim @luispadron I think we are all set to create release 1.0.0. Is there anything else that we need to consider before pulling the trigger?

**Comment by luispadron** [comment]
Only thing I can think of is if fixing https://github.com/cgrindel/rules_swift_package_manager/issues/892 would introduce any breaking changes, maybe we'd want to wait for that? But I think changes should be contained to the upstream rules

**Comment by cgrindel** [comment]
@luispadron TBH, I have not been following that issue very closely. It sounds like the gist is Xcode allows building different packages with different flags, while Bazel's rules, by default, want to build everything with the same set of flags. I can see imagine that `rules_swift_package_manager` would need to generate build files for the packages differently, but I don't think there would be an API/usage change. Let me know if I am missing something.

**Comment by AttilaTheFun** [comment]
@cgrindel I think we should not cut 1.0 without this and Xcode 16 / Swift 6 support:
https://github.com/cgrindel/rules_swift_package_manager/pull/1423

There are popular SPM packages like grpc-swift using Swift 6 features that will fail with RSPM.

Additionally, can we break the dependencies on gazelle, swift_gazelle_plugin and rules_go for the module, even if they're used in some examples?

**Comment by cgrindel** [comment]
Swift 6 support was never part of the release 1.0.0 plan. Adding support should not be a breaking change. 

With regard to gazelle dependencies in the examples, I have no intention of removing it there. They help with the maintenance of the examples. 

With regard to Go deps in rspm, if someone wants to reimplement the Swift deps tool in Swift, that is fine with me. However, I don't anticipate working on it at any time soon. Since it will not be a breaking change for the ruleset, I don't see it as a blocker for release 1.0.0.

**Comment by cgrindel** [comment]
> There are popular SPM packages like grpc-swift using Swift 6 features that will fail with RSPM.

@AttilaTheFun What is needed to support Swift 6?

**Comment by AttilaTheFun** [comment]
@cgrindel There are several SwiftSettings introduced in Swift 6 for enabling various language features like private imports, strict concurrency checks, requiring existential any etc. RSPM will fail on packages that use them.

Re: go deps, it's fine for them to be dev dependencies but IMHO it would be a surprise to me as a developer if a repository rule for swift packages added a dependency on rules_go for no apparent reason.

With gazelle it makes sense because it's a go tool, but for people just using RSPM they shouldn't need it.

**Comment by luispadron** [comment]
> Re: go deps, it's fine for them to be dev dependencies but IMHO it would be a surprise to me as a developer if a repository rule for swift packages added a dependency on rules_go for no apparent reason.

It shouldn't unless you're using the `swift_deps_info` stuff

**Comment by luispadron** [comment]
But to Chucks point, i dont think adding Swift 6 support changes anything about the public API for rules_swift_package_manager and so its not really a blocker for 1.0.0

**Comment by AttilaTheFun** [comment]
@luispadron is it smart enough to not load the go dependencies if the module depends on them but the specific rule does not? At the very least it could cause versioning issues.

@cgrindel not sure about your feelings regarding AI tools but I have found Claude to be very good at translating things between languages. I'd be happy to try feeding the deps tool into it and asking it to translate it to Swift + fixing it up.

**Comment by cgrindel** [comment]
> Re: go deps, it's fine for them to be dev dependencies but IMHO it would be a surprise to me as a developer if a repository rule for swift packages added a dependency on rules_go for no apparent reason.

[Some details](https://bazelbuild.slack.com/archives/C05NZSHE7FU/p1742485749556829?thread_ts=1742474820.093049&cid=C05NZSHE7FU) from @fmeum:

> What you are seeing isn't the actual Go SDKs, it's just the small repo with their toolchain targets
> That repo is fetched eagerly with any dep on rules_go, but it's just a single file

tl;dr It is not downloading the Go toolchain implementation unless you enable the swift deps info functionality.

**Comment by cgrindel** [comment]
> @cgrindel not sure about your feelings regarding AI tools but I have found Claude to be very good at translating things between languages. I'd be happy to try feeding the deps tool into it and asking it to translate it to Swift + fixing it up.

I am happy to review. However, I still do not believe that this is required for release 1.0.0. Again, swapping out the underlying tool should not be a breaking change. 

Keep in mind, you are only enabling this functionality, if you want to use the Swift gazelle plugin. Using gazelle will bring down the Go toolchain among other things.

**Comment by cgrindel** [comment]
> With gazelle it makes sense because it's a go tool, but for people just using RSPM they shouldn't need it.

Much of the work that went into release 1.0.0 was mitigating the Go dependencies. We have it squirreled away such that you don't take the hit unless you want that functionality.

**Comment by AttilaTheFun** [comment]
Ok that seems reasonable though I would still like to help break the non-dev go dependencies at some point. People might see them on registry.bazel.build, not know how or when they're used, and be confused.

**Comment by AttilaTheFun** [comment]
Regarding the SwiftSettings, I understand it's not a breaking change, but I still think announcing the 1.0 release while not supporting an increasing number of packages is strange.

**Comment by luispadron** [comment]
That's fair, what's left for swift 6 support?

**Comment by AttilaTheFun** [comment]
@luispadron I have a pr for the swift settings but it's blocked on the CI running Xcode 15. @cgrindel said there were some errors when he tried to upgrade but we haven't had time to debug them yet.

**Comment by cgrindel** [comment]
@AttilaTheFun So, we need to drop support for Xcode 15?

**Comment by AttilaTheFun** [comment]
@cgrindel I don't know if we have to drop support for Xcode 15, but some packages like grpc-swift now require Swift 6. Specifically I believe it was my new example exercising these Swift 6 features that was failing in CI

**Comment by cgrindel** [comment]
@AttilaTheFun Do you think that merging #1423 is sufficient to declare victory for release 1.0.0?

**Comment by AttilaTheFun** [comment]
@cgrindel yeah I think so -- I just rebased it. It would be nice to mark rules_go as a dev_dependency but I suppose that's not a hard blocker. Not sure if removing it later would be considered a breaking change @brentleyjones ?

**Comment by cgrindel** [comment]
Let's create a release candidate (e.g., 1.0.0-rc1). It will signal our intention to create a 1.0.0 release. It may encourage folks to try it out before we declare victory.

**Comment by cgrindel** [comment]
[Release 1.0.0](https://github.com/cgrindel/rules_swift_package_manager/releases/tag/v1.0.0)
